straxen/plugins/peaks/peak_subtyping.py

```diff
@@ -0,0 +1,621 @@
+    # because we need to make sure the returned containers are not overlapping
+    # prevent numpy change the dtype to float
+    # limitation needs also be sorted
+    if len(limitation) > 0:
+        limitation = np.unique(np.hstack([anchor, limitation]))
+    else:
+        limitation = anchor.copy()
+    containers = np.zeros(len(anchor), dtype=strax.time_fields)
+    if forward:
+        containers['time'] = anchor
+        containers['endtime'] = _limited_containers_forward(anchor, length, limitation)
+    else:
+        containers['endtime'] = anchor
+        containers['time'] = _limited_containers_reverse(anchor, length, limitation)
+    assert np.all(containers['time'][1:] - containers['endtime'][:-1] >= 0)
+    return containers
+
+
+@numba.njit
+def _limited_containers_forward(time, length, limitation):
+    """
+    Construct the endtime in jit accelerated way
+    """
+    endtime = time + length
+    n = len(limitation)
+    i = 0
+    for j in range(len(time)):
+        while i < n - 1 and time[j] >= limitation[i]:
+            i += 1
+        endtime[j] = min(endtime[j], limitation[i])
+    return endtime
+
+
+@numba.njit
+def _limited_containers_reverse(endtime, length, limitation):
+    """
+    Construct the time in jit accelerated way
+    """
+    time = endtime - length
+    n = len(limitation)
+    i = n - 1
+    for j in np.arange(len(endtime))[::-1]:
+        while i > 0 and endtime[j] <= limitation[i]:
+            i -= 1
+        time[j] = max(time[j], limitation[i])
+    return time
+
+
+def combine_indices(result):
+    """
+    Combine the indices from touching_windows results
+    :param result: touching_windows results, each row is a pair of indices
+    """
+    if len(result) == 0:
+        return np.array([], dtype=np.int)
+    indices = []
+    for r in result:
+        indices.append(np.arange(r[0], r[1]))
+    indices = np.unique(np.hstack(indices))
+    return indices
+
+
+def combine_indices_ref(result, areas, reference_areas):
+    """
+    Combine the indices from touching_windows results, based on the areas.
+    If the areas larger than the reference areas, combine them into large_indices,
+    if not, combine them into small_indices.
+    :param result: touching_windows results, each row is a pair of indices
+    :param areas: areas of the peaks of the cooresponding indices
+    :param reference_areas: reference areas to compare with for each pair of indices
+    """
+    if len(result) == 0:
+        return np.array([], dtype=np.int), np.array([], dtype=np.int)
+    if not hasattr(reference_areas, '__len__'):
+        reference_areas = np.full(len(result), reference_areas)
+    assert len(result) == len(reference_areas), ''\
+        + 'result and reference_areas must have the same length'
+    small_indices = []
+    large_indices = []
+    for r, ref in zip(result, reference_areas):
+        small_indices.append(np.arange(r[0], r[1])[areas[r[0]:r[1]] <= ref])
```

Comment:
I would also break this down to make it more readable, but also a bit more efficient e.g.: 

```
_area_is_larger_reference = areas[...] > ref
.... np.arange(r[0], r[1])[_area_is_larger_reference]
.... np.arange(r[0], r[1])[~_area_is_larger_reference]
```